pilot/pkg/proxy/envoy/v2/ads.go

```diff
@@ -263,6 +289,21 @@ func (s *DiscoveryServer) StreamAggregatedResources(stream ads.AggregatedDiscove
 	}
 }
 
+// handleTypeURL records the type url received in an XDS response. If this conflicts with a previously sent type,
+// an error is returned. For example, if a v2 cluster request was sent initially, then a v3 response was received, we will throw an error.
+// This is to ensure that when we do pushes, we are sending a consistent type, rather than flipping between v2 and v3.
+// A proper XDS client will not send mixed versions.
```

Comment:
@htuch is this a safe assumption? I am realize now it could be wrong if we send a CDS response that references a v2 EDS in one cluster and v3 EDS in another cluster, but assuming we ensure we are consistent in our response?

Reply:
So "a proper XDS client" is probably wrong here, more accurately "an XDS client configured in the way we configure it"

Reply:
I think the CDS response can have clusterA reference a v3 EDS resource and clusterB reference a v2 EDS response. You would end up with two distinct DiscoveryRequess on an ADS channel for the distinct resource type URLs...

Reply:
@htuch Is this possible the requests can be different versions from a same connection?

Reply:
Yes, as described above - each cluster references an eds version which could be different. In practice we will ensure they are in sync